### PR TITLE
[SPARK-33079][TESTS] Replace the existing Maven job for Scala 2.13 in Github Actions with SBT job

### PR DESCRIPTION
**Pull request by sarutak**
### What changes were proposed in this pull request?

SPARK-32926 added a build test to GitHub Action for Scala 2.13 but it's only with Maven.
As SPARK-32873 reported, some compilation error happens only with SBT so I think we need to add another build test to GitHub Action for SBT.
Unfortunately, we don't have abundant resources for GitHub Actions so instead of just adding the new SBT job, let's replace the existing Maven job with the new SBT job for Scala 2.13.

### Why are the changes needed?

To ensure build test passes even with SBT for Scala 2.13.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

GitHub Actions' job.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
cc @HyukjinKwon

File: .github/workflows/build_and_test.yml
Comment (HyukjinKwon) [inline .github/workflows/build_and_test.yml:323]: @sarutak, can we actually fix the main job https://github.com/apache/spark/blob/5dec1ffc9eaae07b6c71ebb2de4b64da87328390/.github/workflows/build_and_test.yml#L16-L80 to run the tests?

File: .github/workflows/build_and_test.yml
Comment (HyukjinKwon) [inline .github/workflows/build_and_test.yml:324]: Hm, should we set this job in the PR builder? The (almost) same thing is tested at "Scala 2.13 build". I think we'll set a Jenkins regular job to run the tests at SPARK-33044 as well. I would avoid adding too many jobs here since we're not able to have more resources (see also SPARK-32264).

Comment (sarutak) [reply .github/workflows/build_and_test.yml:324]: > The (almost) same thing is tested at "Scala 2.13 build"

The main purpose of adding the build with SBT is not just testing with SBT. As I mentioned, some compilation error happens only with SBT so I'd like detect such error before any changes being merged. 
I understand we have not enough resource for GitHub Action but if an additional build is added to the Jenkins regular job, it will take much more longer time doesn't it?
Are there any problem to replace the existing "Scala 2.13 build" with the new one?

Comment (HyukjinKwon) [reply .github/workflows/build_and_test.yml:324]: I am okay with replacing the existing "Scala 2.13 build". Seems that'll catch the compilation errors more correctly. 

> it will take much more longer time doesn't it?

It does takes longer to detect the errors. However, there are many other combinations to test and currently we can't test all combinations. So I think the default common-ish combination was picked to use it as a GitHub Actions and Jenkins PR builder. For example, JDK 11 is not being tested in GitHub Actions and Jenkins PR builder.

Comment (sarutak) [reply .github/workflows/build_and_test.yml:324]: > For example, JDK 11 is not being tested in GitHub Actions and Jenkins PR builder.

Yeah, I know but whether build is successfully finish is tested in GitHub Actions so I thought build test (not unit test) should be checked in GitHub Actions too for now.
Anyway, If we can replace the existing "Scala 2.13 build" with the new one, I'll do it.
@dongjoon-hyun Do you have any concern about replacing it?

File: external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisBackedBlockRDD.scala
Comment (LuciferYang) [inline external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisBackedBlockRDD.scala:94]: Is this change related to the current pr?

Comment (sarutak) [reply external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisBackedBlockRDD.scala:94]: Yes. empty parameter list causes compilation error if we build for Scala 2.13 with SBT.
This is what I'd like to detect by CI.

File: .github/workflows/build_and_test.yml
Comment (HyukjinKwon) [inline .github/workflows/build_and_test.yml:297]: I would remove -sbt postfix

File: .github/workflows/build_and_test.yml
Comment (HyukjinKwon) [inline .github/workflows/build_and_test.yml:298]: and here too

File: .github/workflows/build_and_test.yml
Comment (HyukjinKwon) [inline .github/workflows/build_and_test.yml:316]: I guess we don't have to do this. The test cases above are using SBT too but it doesn't create the directory.

File: .github/workflows/build_and_test.yml
Comment (HyukjinKwon) [inline .github/workflows/build_and_test.yml:317]: I think we can remove `mkdir -p ~/.m2` here and `rm -rf ~/.m2/repository/org/apache/spark` as we don't cache `.m2` anyway here.

**Review by HyukjinKwon — APPROVED** [review]
I am good with this change.

File: .github/workflows/build_and_test.yml
Comment (sarutak) [reply .github/workflows/build_and_test.yml:317]: `.m2` is required because `publishLocal` installs artifacts into `.m2`.
But, as I mentioned above, I think we can only do `compile` and `test:compile` so I'll remove this part.

**Comment by LuciferYang** [comment]
@sarutak It seems that the title and description of this pr need to be modified

**Comment by sarutak** [comment]
> @sarutak It seems that the title and description of this pr need to be modified
Done.

**Comment by HyukjinKwon** [comment]
I'll merge this after leaving it some more days in case people have different thoughts.

File: .github/workflows/build_and_test.yml
Comment (dongjoon-hyun) [inline .github/workflows/build_and_test.yml:317]: So, we are going to remove `packaging` and `installation (Maven)` here?

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM for this approach. Thank you, @sarutak and @HyukjinKwon .
I'll leave this PR to you guys~

File: .github/workflows/build_and_test.yml
Comment (LuciferYang) [reply .github/workflows/build_and_test.yml:317]: Can Maven build and SBT build be executed serially in one action？

Comment (HyukjinKwon) [reply .github/workflows/build_and_test.yml:317]: I think it wouldn't matter too much. If a job is finished quicker the next job can be triggered

Comment (LuciferYang) [reply .github/workflows/build_and_test.yml:317]: It seems that 2.13 build action runs faster than other actions. Can we consider this ?

Comment (HyukjinKwon) [reply .github/workflows/build_and_test.yml:317]: The number of job is limited in github actions instead of each entire workload run IIRC.

**Comment by srowen** [comment]
Can we keep Maven and SBT tests for 2.13? Maven is the build of reference anyway

**Comment by HyukjinKwon** [comment]
@srowen, maybe i think we'll have to setup the regular Jenkins job for that like we have done for other profiles.  I agree that it's best to keep both Maven and SBT build in Scala 2.13 (and actually we should ideally test all combinations like JDK 11 with running the tests) but the problem is limited resource. The resource is currently all shared in all ASF projects although ASF has an Enterprise plan.

Maybe we could focus on PR builders to catch the errors with a minimized set. For reference or other stuff, it might be better to set these jobs in regular Jenkins jobs.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**
